Automatically determine a more optimal -dbcache setting if none provided

ptschip opened the pull request:
In the past we always defaulted to a 500MB value for -dbcache if the node operator didn't provide one however this value would still cause a node to take up to 3 weeks or more to sync and is thus inadequate. The problem remains in that node operators, both new and experienced, still forget to set the dbcache to it's highest possible before doing an initial sync, wasting valuable time and also developer time in fielding numerous complaints of slow sync.  This PR attempts to remedy that by setting the -dbache to one half the physical memory size "IF" the node operator has not already defined a value in the bitcoin.conf or on startup through the command line.

@gandrewstone can you test on  OSX and verfy it's actually picking up the correct values for physical RAM...thanks.  UNIX and Windows are verified to be working.

ptschip committed:
Kill QT dbcache setting in Options

This setting is of little use since the operator has to restart
the application anyway to make the setting take effect.  By removing
this and only using -dbcache from either the command line or bitcoin.conf
it gives us the ability to automatically configure the dbcache and
also simplifies the configuration process.

ptschip committed:
Add functions to determine the max system memory and use them

Use half the total system memory as the value for -dbcache
and use this value if the node operator has not already set
a nummber for -dbcache.  This way we can be pretty sure a node
will get a much better initial sync rather than in the past where
we would only rely on the default of 500MB for -dbcache.

Add getTotalSystemMemory() for UNIX and MACOS

Also return with a default value if physical memory size could
not be determined. We return with a default of 1GB which should
cover any raspberry pi configurations.

Make a more accurate calculation on Windows for memory available

On Windows systems we can know how much of the physical RAM is available
for use and so here we make a more fine grained calculation for the
amount of memory we'll use for the dbcache mechanism. For Windows systems
with large amounts of RAM available we'll use all of it up to the 16GB
maximum allowed for dbcaching, however, we'll always leave 1GB free
to the system for other uses.

ptschip committed:
Use an int64_t instead of size_t for nCoinCacheUsage

Because nCoinCacheUsage is size_t, then on 32bit systems when
we assign it a value from 64bit int then we end up with an
incorrect value. This happens during init when the cache sizes
are being calculated.

ptschip committed:
Move memory functions to memusage.h in preparation for dynamic dbcache sizing on Windows

ptschip committed:
On Windows, dynamically reduce the size of the dbcache if/when needed.

If the available memory falls below 10% then
reduce the size of the dbcache's coin cache.

ptschip committed:
Allow the retreival of the lowest default dbcache configuration

This is needed to prevent the coins cache from getting too low
during the process of dynamically changing the dbcache settings.

ptschip committed:
On Windows, dynamically increase the size of the dbcache if possible.

ptschip committed:
Cleanup FlushStateToDisk by moving the dynamic coin cache adjustment

Create AdjustCoinCacheSize() and use it in FlushStateToDisk and move
the underlying code to txdb.cpp.

ptschip committed:
change from LOGA to LOG(COINDB, ...

ptschip committed:
Update README.md with instruction on setting up for initial sync.

Describe the importance of the dbcache setting in the initial sync
process. Also touch upon the automatic dbcache configuration as well
as the possible need to set a higher maxoutconnections.

ptschip committed:
Add comments for thew new functions in txdb.h

ptschip commented:
This is ready for testing if anybody wants to try it out.  @Justaphf  @Greg-Griffith , I know you guys run on Windows and this PR works especially well on Windows.  I've tested it out and watched how it sizes the dbcache and if you launch other apps which take up too much mem, then the dbcache gets floated down in size...then later if you kill those other apps the dbcache will rise again.  It works AFAIK but it would be great to get others trying it out.  As for Linux or Mac I haven't tested on those...

gandrewstone commented on src/txdb.cpp:
```diff
@@ -428,3 +428,210 @@ bool CCoinsViewDB::Upgrade()
+    int64_t nPhysMem = 0;
+    size_t nLength = sizeof(nPhysMem);
+
+    if (sysctl(mib, 2, &nPhysMem, &nLength, nullptr, 0))
```
tested on apple.  This line needs to be:
if (sysctl(mib, 2, &nPhysMem, &nLength, nullptr, 0)==0)

ptschip replied:
@gandrewstone thanks, done...

gandrewstone reviewed (changes requested):
small mac fix

ptschip committed:
Fix GetTotalSystemMemory for MacOS

a zero indicates a positive return value for sysctl rather
than a 1

ptschip committed:
Use a better size estimate when determining when to batch write

Since the time of the datbase upgrade the batch size was not
being calculated properly and as a result we were not batching
our writes anymore. Here we get the appropriate batch size
estimate and write our batches accordingly.

ptschip committed:
On Windows, ensure that there is enough free memory for the write caches

On very marginal nodes we have to at least leave enough memory for the
write caches and also be careful to trim the coins cache more
carefully.

ptschip committed:
fix compiler warnings

ptschip commented:
@gandrewstone I've tested this on a variety of windows systems, 32 bit 64 bit  , small RAM and beefier systems.  The only trouble I had was on the marginal RAM 32bit system which  I fixed with 7a2e564 and 57b82f7 .

sickpig commented on README.md:
````diff
@@ -37,7 +37,7 @@ If you're running an Ubuntu system:
 
 ```sh
 sudo apt-get install software-properties-common
-sudo add-apt-repository ppa:bitcoin-unlimited/bucash
````
this change and the next one seem unrelated to this PR but I'm fine to getting that merged  via this PR

ptschip replied:
i''m not sure how that even got in there...that wasn't from me...strange.  I must have grabbed a more recent readme file from somewhere else and used it...happy it's the right one though.

sickpig commented on README.md:
```diff
@@ -68,3 +68,31 @@ sudo make install
+
+### QT or the command line:
+
+There are two modes of operation, one uses the QT UI and the other runs as a daemon from the command line.  The QT version is bitcoin-qt or bitcoin-qt.exe, the command line version is bitcoind or bitcoind.exe. No matter which version you run, when you launch for the first time you will have to complete the intial blockchain sync.
```
s/intial/initial

ptschip committed:
Once IBD is complete then trim the coins cache

If the user has not specifiied a dbcache setting then trim
the coins cache to the default size once IBD is completed.

ptschip committed:
Flush state and trim coins cache before shutting down

At the beginning of the shutdown process, flush and clear the coins
cache to free up as much memory as possible. This helps to prevent
any runaway exceptions due to a bad alloc during the shutdown process.

ptschip committed:
Keep 300MB available for marginal Windows systems
